### ISSUE TITLE
Chrome Extensions support?

### ISSUE THREAD

**Issue by MrHayato** [systemjs/systemjs#421]
I've been using jspm in most of my projects, and now I'm looking to add it to a new project for a Chrome Extension I'm working on.

One of the limitations of writing Chrome Extensions, however, is that as of manifest version 2, [Inline scripts are no longer supported](https://developer.chrome.com/extensions/tut_migration_to_manifest_v2#security-summary). Thus, I am unable to load in my scripts dynamically using system.js.

I have managed to get around this by modifying system.src.js's `doEval` function, to simply eval:

```javascript
doEval = function(source) {
   eval(source);
};
```

This works fine now, and I am able to load in my scripts from jspm from my extension as well as as a web app.

Another possibility, is to change `script.text = source` to be `script.src = address` but then it didn't execute the contents immediately, which would require a bunch of callbacks.

Is there another way we could potentially load in the scripts without inline-ing them in a script tag so we could support chrome extensions? Or perhaps a flag to do it another way?

**Comment by matthewp** [comment]
There is a csp build for this: https://github.com/systemjs/systemjs/tree/master/dist

**Comment by MrHayato** [comment]
Oh, perfect. Is there a setting in jspm to use the csp version of system.js for fresh jspm installs?

**Comment by matthewp** [comment]
I'm not familiar with jspm but in your script tag you just change it to system.csp.js I would imagine.

**Comment by MrHayato** [comment]
Yeah I can do that in my index.html, but when I run `jspm install` it brings in system.js with it. For deployments and my machine I can manually change it, but for other developers they would need to remember to replace it with the csp version.

**Comment by matthewp** [comment]
I don't follow, why can't you commit your change so other developers get it?

**Comment by MrHayato** [comment]
jspm is like bower, and brings in packages. jspm also brings in system.js (just system.js, system.src.js, and the sourcemaps, not the whole repository) automatically. I .gitignore the jspm_packages folder (where jspm installs all the packages) because those get generated when the developer runs `jspm install`. I could commit the `jspm_packages/system.js` and ignore the rest of the files in the folder, but I'd rather not do that if there's a way to get the system.csp.js file instead.

**Comment by MrHayato** [comment]
Hmmm, csp version of systemjs doesn't seem to work either. I keep getting `Multiple anonymous System.register calls in the same module file.` FWIW I'm trying to get an [Aurelia](http://aurelia.io) app running in a chrome extension.

**Comment by MajorBreakfast** [comment]
In order for the csp version to work you need to build or bundle your app with jspm.

Also, maybe you could open an issue on the jspm repo about including the csp build in the packages folder automatically. That way more people would know about it and using it becomes more straightforward.

**Comment by guybedford** [comment]
This is a bug - we should be detecting the chrome extension environment and using eval over script injection in this case.

The specific code would go in https://github.com/systemjs/systemjs/blob/master/lib/polyfill-wrapper-end.js#L26 where we determine the right evaluation method for the environment.

PRs welcome.

**Comment by guybedford** [comment]
Any ideas of what environment detection code to use here?

**Comment by guybedford** [comment]
I've added this in https://github.com/systemjs/systemjs/commit/29cd262a386d2c978696941601f0979c28c35350 - let me know if that seems right!

**Comment by matthewp** [comment]
I thought chrome extensions couldn't do eval at all. Am I wrong?

**Comment by guybedford** [comment]
I guess it can run inside a sandbox - https://developer.chrome.com/extensions/sandboxingEval, which must be what @MrHayato is doing? I'm not that familiar with the environment so if there are better approaches please let me know.

**Comment by guybedford** [comment]
I've included the possible fix in the current release. Still very experimental if this actually works - feedback welcome.

**Comment by ebaranov** [comment]
@guybedford looks like it didn't work as expected. I didn't test the bundled version, but with the fix 29cd262a386d2c978696941601f0979c28c35350 we totally disabled possibilities to test extensions in AMD mode. Currently in my google chrome extension I've disabled the checking from commit 29cd262a386d2c978696941601f0979c28c35350, and I'm able to use AMD for my extension's options page and content scripts. As for content scripts the Google SCP allow us to configure unsafe eval via manifest https://developer.chrome.com/extensions/contentSecurityPolicy#relaxing-eval

I'll go forward to test bundled version of extension with latest fix 29cd262a386d2c978696941601f0979c28c35350

**Comment by ebaranov** [comment]
So, I tested in google chrome and it works for me only using the ``bundle-sfx``, e.g. 
``jspm bundle-sfx my_main_script out.js``
it works fine for me. 

P.S. I'm using about 15 jspm pckages and about 40 internal modules, it's about 120 es6 files with deep dependencies.

**Comment by MajorBreakfast** [comment]
I dabbled with chrome web apps once (same csp AFAIK) and if I remember correctly the normal systemjs worked in sandbox mode just fine. The point of the csp version is to avoid that mode and its restrictions.

**Comment by ebaranov** [comment]
@MajorBreakfast can't imagine how to migrate a lot of existing code to sandbox mode :)

BTW requirejs is working well for chrome extension.

@guybedford did you look at requirejs injection implementation? Maybe it's possible to borrow theirs approach...

**Comment by yoavniran** [comment]
Hi, Im new to systemjs and jspm but wanted to give it a try for a new chrome extension im working on. So... installed jspm which uses systemjs.

I added refs to system.js and  config.js to my html page and reloaded the extension. Nothing works and I get this error in the console: **Uncaught ReferenceError: System is not defined**

I tried the system-csp.js version but same result. 

Debugging a bit I can see that the only lines of code that actually get executed are: 
```js
if ($__global.chrome && chrome.extension) {
    doEval = 0 || eval; // for uglify
  }
```

If I just load the html page in the browser, it seems to work.

Guessing I'm doing something wrong but I cant figure out what...

**Comment by guybedford** [comment]
Ok I've reverted the Chome evaluation exception in https://github.com/systemjs/systemjs/commit/08b227a4f84b03ea3e9d50d47fcec4bb85268375 for the next release. I guess the point is that it will work with CSP explicitly disabled, otherwise use SystemJS-CSP.

@yoavniran you probably want to ensure you're loading es6-module-loader.js before system.js.

**Comment by yoavniran** [comment]
thanks @guybedford 
forgot to mention that I tried that but then System.config doesnt exist.
it looks to me that system.js (version 0.16.9) doesnt actually work within a chrome extension...

**Comment by yoavniran** [comment]
@guybedford 
with the fix you mentioned you committed (08b227a) I was able to get it working. 
thanks

**Comment by jaysonminard** [comment]
@MrHayato I got background pages and browser action working fine using the system-csp version, did you load modules for content scripts, or do something else?  content scripts want to load from the domain of the page they are injected into, rather than from the plugin.  And specifying the full URL within the plugin doesn't work either (says that System is not defined).

**Comment by jaysonminard** [comment]
As far as I can get with unbundled content scripts is to change baseUrl to the extension root just for content scripts, but then run into "Uncaught ReferenceError: System is not defined" in the included script, whereas the other scripts being used in the content script see it fine.  So loading:

```
"js": [
        "patch/system-csp-production-0.18.2-dev.js",
        "preconfig-inject.js",
        "config.js",
        "pages/inject/content-script-test.js"
      ]
```

with preconfig-inject.js setting the baseUrl based on the extension ID, and the injected `content-script-test.js` being a simple:

```
System.import("inject/google").then(function(m) {
    m.cat();
});
```

for now I'll just bundle because I do very little with system.js but bring in one or two dependencies to content scripts.

**Comment by guybedford** [comment]
@jaysonminard thanks for the update here. Can you clarify which script in particular is giving System is not defined? Does it seem like the baseURI is being detected incorrectly then do you think?

**Comment by jaysonminard** [comment]
Yes, baseURL is ok, it finds the script to import, reads it, and that script has System undefined.  So inject/google calls System.register and that is the line that fails. The prior scripts are loaded by the manifest for script injection, and then baseUrl set to a URL in the chrome extension (which is different than how the other scripts are loaded), then it loads and doesn't have access to things from the prior scripts in the way it is loaded.

**Comment by jaysonminard** [comment]
#531 appears to resolve this.  Testing more.

**Comment by jaysonminard** [comment]
So this is working fine for loading scripts in all aspects of a chrome plugin (background page, browser action, and content scripts).  For background page, it is pretty much normal usage, nothing fancy.  For browser action and content scripts, before config.js is loaded, I have a preconfig.js that contains:

```
System.config({
    baseURL: chrome.extension.getURL("/")
});
```

And then make sure I allow web accessible resources for the directories it will try to access.

**Comment by guybedford** [comment]
Excellent, thanks. If there is anything else that would help just let me know.

**Comment by arogg** [comment]
OK getting grey hair from the last problem I'm having because I can't find a solution to the same problem which was mentioned here:

https://github.com/systemjs/systemjs/issues/421#issuecomment-114420837

The problem I am having is the same problem mentioned in this post, but although I am using 0.18.17 (the latest), the problem that was discussed still exists (in my project at least) and does not seem to have been fixed:

[app.zip | uploaded via ZenHub](https://files.zenhub.io/55f967710f3fd68516842076)

In the content scripts the file "content2.js" imports "dummyinclude.js". But in "dummyinclude.js" it tells me that System is undefined.

I would be grateful to anyone that could inspect the content_scripts part of my manifest and everything that has to do with it for any errors I made. That would be a real help to me.

Thanks for anyone that is willing to help me out on this!
